core-libs/setup/ssr/engine/ng-express-engine.ts

```diff
@@ -0,0 +1,122 @@
+/*
+ * SPDX-FileCopyrightText: 2023 SAP Spartacus team <spartacus-team@sap.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { StaticProvider } from '@angular/core';
+import {
+  CommonEngine,
+  CommonEngineOptions,
+  CommonEngineRenderOptions,
+} from '@angular/ssr';
+import { Request, Response } from 'express';
+import { REQUEST, RESPONSE } from '../tokens/express.tokens';
+
+/**
+ * @license
+ * The MIT License
+ * Copyright (c) 2010-2023 Google LLC. http://angular.io/license
+ *
+ * See:
+ * - https://github.com/angular/universal/blob/e798d256de5e4377b704e63d993dc56ea35df97d/modules/express-engine/src/main.ts
+ */
+export type NgSetupOptions = Pick<
+  CommonEngineRenderOptions,
+  'providers' | 'publicPath' | 'inlineCriticalCss'
+> &
+  CommonEngineOptions;
+
+/**
+ * @license
+ * The MIT License
+ * Copyright (c) 2010-2023 Google LLC. http://angular.io/license
+ *
+ * See:
+ * - https://github.com/angular/universal/blob/e798d256de5e4377b704e63d993dc56ea35df97d/modules/express-engine/src/main.ts
+ */
+function getReqResProviders(req: Request, res?: Response): StaticProvider[] {
+  const providers: StaticProvider[] = [
+    {
+      provide: REQUEST,
+      useValue: req,
+    },
+  ];
+  if (res) {
+    providers.push({
+      provide: RESPONSE,
+      useValue: res,
+    });
+  }
+
+  return providers;
+}
+
+/**
+ * @license
+ * The MIT License
+ * Copyright (c) 2010-2023 Google LLC. http://angular.io/license
+ *
+ * See:
+ * - https://github.com/angular/universal/blob/e798d256de5e4377b704e63d993dc56ea35df97d/modules/express-engine/src/main.ts
+ */
+export interface RenderOptions extends CommonEngineRenderOptions {
+  req: Request;
+  res?: Response;
+}
+
+/**
+ * @license
```

Comment:
Before the @license, let's add to this JSDoc information for us and for customers - why we need it. That it was present in Angular for long time and it got removed only in version Angular 17. But we neeed it in Spartacus, so we copied their code to our repo for backward compatibility to avoid bigger refactors in Spartacus